internal/gps/manager_test.go

```diff
@@ -488,23 +544,56 @@ func TestGetSources(t *testing.T) {
 					t.Error("explicit https source should reuse autodetected https source")
 				}
 
-				// Now put in http, and they should differ
 				lpi.Source = "http://" + string(lpi.ProjectRoot)
 				srcg4, err := sm.srcCoord.getSourceGatewayFor(ctx, lpi)
 				if err != nil {
 					t.Errorf("unexpected error getting explicit http source: %s", err)
-				} else if srcg == srcg4 {
-					t.Error("explicit http source should create a new src")
+				} else if srcg != srcg4 {
+					t.Error("explicit http source should reuse autodetected https source")
```

Comment:
What exactly are those rules?  Since the map was keyed on both name and url, I thought it was in line with existing behavior to fold them all together.  Was sharing a map only valid since the url was static?

Reply:
so the "rule" here - no, not expressed anywhere, because i have, as we know, not been great at articulating these really low-level invariants anywhere, 💯 my fault - is that if an explicit request is made for e.g. an `http` protocol access pattern for a URL, then we have to respect it, and that's how we access. it's only in the default inference case, where no protocol is provided (a bare import path) where we can "pick" from the various possibilities, and where the specific one used may vary from one run to the next (but, once selected, is fixed within the duration of any given `SourceMgr`'s lifetime).

this might seem to push us in a direction that i was considering (and i think we discussed somehwere?), where the `sourceGateway` effectively becomes more of a factory/intermediate layer that hangs on to the entire `maybeSources` set and multiplexes to children that are responsible for each individual URL, depending on which particular input pattern comes in.

(if we're thinking about that, one crucial reminder about the `maybeSources` model is that e.g. bitbucket abstracts `maybeSources` to allow for a project to be _either_ git or hg - so we're encompassing in there the idea that ).